### DISCUSSION TITLE
How do you use Prisma and SQLite?

### DISCUSSION THREAD

**Discussion by Carw** [cawa-93/vite-electron-builder#640] [Q&A] [answered]
Prisma is simple and useful ORM. But i have a lot of problem with it. Have you tried to use it in your project?

What problems do i have now:

1) I have an error "PrismaClient is unable to be run in the browser" with using `import {PrismaClient} from "@prisma/client"`. It relate with strange import of Prisma.

I found this solve with using require:

```
import {createRequire} from 'module';
const require = createRequire(import.meta.url);
const {PrismaClient} = require('@prisma/client');

export const db = new PrismaClient();
```
It works.

2) Now, when i compile a product, it doesn't work, because can't find Prisma:

> Error: Cannot find module '.prisma/client/index'

Looking for a solution.

3) I need to migrate DB when my app starts. But Prisma have API to  migrate DB only in CLI.

Looking for a solution.

So, how do you use Prisma?

**Comment by cawa-93** [comment, marked as answer]
Binary dependencies (I assume yours is just that) can't be bundled. They must be external.
https://github.com/cawa-93/vite-electron-builder#working-with-dependencies

**Reply by Carw** [reply]
@cawa-93 Thanks for your answer. I use Prisma only for the main process. I think it's bad idea to use it in the renderer. Is it relevant for the main process?

**Reply by cawa-93** [reply]
Yes, because the dependencies in main process is also bundled. The only difference is that you do not need a preload, as the main process has access to the Node.js API in runtime.

**Reply by cawa-93** [reply]
🤔 This is a good remark. I need to think about how to improve the documentation

**Reply by Carw** [reply]
It will be helpful. Modern programs often need a database and data processing. And being able to use Prism and other solutions are very useful. Thanks for the template, it's awesome.

**Reply by vdawg-git** [reply]
I am using Prisma with SQLite and I just put the database in the `public` folder. On app startup it gets copied into the `userData` path and I launch Prisma with the copied one

**Comment by vdawg-git** [comment]
Edit: Check out this template to get around the difficulties of Electron and Prisma: https://github.com/awohletz/electron-prisma-trpc-example/

I am using Prisma with SQLite and I just put the database in the public folder. On app startup it gets copied into the userData path and I launch Prisma with the copied one

(Answering here so that others can see it better)

```javascript
// The "adapted" Prisma client

import { PrismaClient } from "@prisma/client"

export const devDBPath = join(app.getPath("userData"), "devDB.db")
export const productionDBPath = join(app.getPath("userData"), "productionDB.db")
export const dbPath = import.meta.env.DEV ? devDBPath : productionDBPath
export const dbURL = "file:" + dbPath

export default function createPrismaClient() {
  const client = new PrismaClient({
    datasources: {
      db: { url: dbURL },
    },
  })
  return client
}
```
```javascript
// The "creation" of the database, run at app startup
// The Prisma schema itself uses the one in "public", which then gets copied over for the app

// Check if database exists. If not copy the empty master to make it available
export async function checkDatabase(dbPath: string) {
  if (!await checkPathAccessible(dbPath)) { 
    if (import.meta.env.DEV) {
      promises
        .copyFile(join(__dirname, "../public/masterDB.db"), dbPath)
        .catch(log.error.red)
    } else {
      promises
        .copyFile(join(__dirname, "masterDB.db"), dbPath)
        .catch(log.error.red)
    }
  }
}

```
